File: 2023/01/day01.py
import sys

def get_input(file):
    return [line.strip() for line in file]

def part1(puzzle_input):
    return sum([firstlast(line) for line in puzzle_input])

def part2(puzzle_input):
    parsed = [
        line.replace('one', 'o1e')\
            .replace('two', 't2o')\
            .replace('three', 't3e')\
            .replace('four', 'f4r')\
            .replace('five', 'f5e')\
            .replace('six', 's6x')\
            .replace('seven', 's7n')\
            .replace('eight', 'e8t')\
            .replace('nine', 'n9n')
        for line in puzzle_input]

    return sum([firstlast(line) for line in parsed])

def firstlast(line):
    digits = [int(c) for c in line if c.isnumeric()]
    return digits[0] * 10 + digits[-1] if len(digits) > 0 else 0

File: 2023/01/day01.ex
defmodule Day01 do
  def get_input(file) do
    file
    |> IO.read(:eof)
    |> String.split("\n", trim: true)
  end

  def part1(input) do
    input
    |> Enum.map(&firstlast/1)
    |> Enum.sum()
  end

  def part2(input) do
    for line <- input do
      line
      |> String.replace("one", "o1e")
      |> String.replace("two", "t2o")
      |> String.replace("three", "t3e")
      |> String.replace("four", "f4r")
      |> String.replace("five", "f5e")
      |> String.replace("six", "s6x")
      |> String.replace("seven", "s7n")
      |> String.replace("eight", "e8t")
      |> String.replace("nine", "n9e")
      |> firstlast()
    end
    |> Enum.sum()
  end

  defp firstlast(s) do
    digits = for c <- String.to_charlist(s), ?1 <= c and c <= ?9, do: c - ?0
    case {List.first(digits), List.last(digits)} do
      {nil, nil} -> 0
      {tens, ones} -> tens * 10 + ones
    end
  end
end
